### PR TITLE
Implement Custom Action Buttons. Stage2

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

File: js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 1]
```diff
@@ -27,7 +27,8 @@ var DROP_DOWN_EDITOR_CLASS = "dx-dropdowneditor",
     DROP_DOWN_EDITOR_OVERLAY = "dx-dropdowneditor-overlay",
     DROP_DOWN_EDITOR_OVERLAY_FLIPPED = "dx-dropdowneditor-overlay-flipped",
     DROP_DOWN_EDITOR_ACTIVE = "dx-dropdowneditor-active",
-    DROP_DOWN_EDITOR_FIELD_CLICKABLE = "dx-dropdowneditor-field-clickable";
+    DROP_DOWN_EDITOR_FIELD_CLICKABLE = "dx-dropdowneditor-field-clickable",
+    DROP_DOWN_EDITOR_FIELD_TEMPLATE_WRAPPER_CLASS = "dx-dropdowneditor-field-template-wrapper";
```
Comment (churkin) [inline js/ui/drop_down_editor/ui.drop_down_editor.js]: extra `_CLASS` postfix

File: js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 2]
```diff
@@ -376,14 +377,16 @@ var DropDownEditor = TextBox.inherit({
         this._disposeKeyboardProcessor();
 
         // NOTE: to prevent buttons disposition
-        this.$beforeButtonsContainer && this.$beforeButtonsContainer[0].parentNode.removeChild(this.$beforeButtonsContainer[0]);
-        this.$afterButtonsContainer && this.$afterButtonsContainer[0].parentNode.removeChild(this.$afterButtonsContainer[0]);
+        this._$beforeButtonsContainer && this._$beforeButtonsContainer[0].parentNode.removeChild(this._$beforeButtonsContainer[0]);
```
Comment (DokaRus) [inline js/ui/drop_down_editor/ui.drop_down_editor.js:380]: Why not
```suggestion
        this._$beforeButtonsContainer.remove();
```
?

File: js/ui/text_box/action_button_collection/button.js [hunk 3]
```diff
@@ -1,14 +1,14 @@
 import $ from "../../../core/renderer";
 
 export default class ActionButton {
-    constructor(name, editor, { location = "after" }) {
+    constructor(name, editor, options) {
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/button.js:4]:
```suggestion
    constructor(name, editor, options = { location = "after" }) {
```
Is default value unnecessary?

File: js/ui/text_box/action_button_collection/index.js [hunk 4]
```diff
@@ -1,7 +1,60 @@
...
+import { extend } from "../../../core/utils/extend";
 import { find } from "../../../core/utils/array";
 
 const TEXTEDITOR_BUTTONS_CONTAINER_CLASS = "dx-texteditor-buttons-container";
+const TEXTEDITOR_BEFORE_BUTTONS_CONTAINER_CLASS = "dx-texteditor-before-buttons-container";
+const TEXTEDITOR_AFTER_BUTTONS_CONTAINER_CLASS = "dx-texteditor-after-buttons-container";
+
+function checkButtonInfo(buttonInfo) {
+    const checkButtonType = () => {
+        if(!buttonInfo || typeof buttonInfo !== "object" || Array.isArray(buttonInfo)) {
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/index.js:10]:
```suggestion
        if(!buttonInfo || !isObject(buttonInfo) || Array.isArray(buttonInfo)) {
```
Use `type` utils

File: js/ui/text_box/action_button_collection/index.js [hunk 5]
```diff
@@ -1,7 +1,60 @@
...
+            throw new TypeError("'buttons' option must include an object or a string items only");
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/index.js:11]: Should we use our `EXXX` errors message with links to the documentation?

File: js/ui/text_box/action_button_collection/index.js [hunk 6]
```diff
@@ -1,7 +1,60 @@
...
+        const { location } = buttonInfo;
+
+        if("location" in buttonInfo && location !== "after" && location !== "before") {
+            throw new TypeError("action button's 'location' property can be 'after' or 'before' only");
+        }
+    };
+
+    const checkNameIsDefined = () => {
+        if(!("name" in buttonInfo)) {
+            throw new Error("action button must have a name");
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/index.js:25]: Same as above

File: js/ui/text_box/action_button_collection/index.js [hunk 7]
```diff
@@ -1,7 +1,60 @@
...
+        }
+    };
+
+    const checkNameIsString = () => {
+        const { name } = buttonInfo;
+
+        if(typeof name !== "string") {
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/index.js:32]: Use `type` utils

File: js/ui/text_box/action_button_collection/index.js [hunk 8]
```diff
@@ -1,7 +1,60 @@
...
+
+    checkButtonType();
+    checkNameIsDefined();
+    checkNameIsString();
+    checkLocation();
+}
+
+function checkNamesUniqueness(existingNames, newName) {
+    if(existingNames.indexOf(newName) !== -1) {
+        throw new Error("'buttons' option item must have unique name");
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/index.js:45]: And here

File: js/ui/text_box/action_button_collection/index.js [hunk 9]
```diff
@@ -10,46 +63,69 @@ export default class ActionButtonCollection {
...
-    _createButton(buttonName) {
-        const defaultButtonInfo = find(this.defaultButtonsInfo, ({ name }) => name === buttonName);
+    _compileButtonInfo(buttons) {
+        const names = [];
 
-        if(!defaultButtonInfo) {
-            throw "Can't create custom action button";
-        }
+        return buttons.map((button) => {
+            const isDefaultButton = typeof button === "string";
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/index.js:68]: Use `type` utils

File: js/ui/text_box/action_button_collection/index.js [hunk 10]
```diff
@@ -10,46 +63,69 @@ export default class ActionButtonCollection {
...
         this.buttons.push(button);
 
         return button;
     }
 
-    _renderButtons(buttons, $container, location) {
-        let $buttonsContainer = $buttonsContainer = $("<div>")
-            .addClass(TEXTEDITOR_BUTTONS_CONTAINER_CLASS)
-            .appendTo($container);
+    _getLocation(buttonInfo) {
```
Comment (DokaRus) [inline js/ui/text_box/action_button_collection/index.js]: Where you use it?

**Review by DokaRus — COMMENTED** [review]
Did you check the RTL?

File: js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 11]
```diff
@@ -376,14 +377,16 @@ var DropDownEditor = TextBox.inherit({
...
+        this._$afterButtonsContainer && this._$afterButtonsContainer[0].parentNode.removeChild(this._$afterButtonsContainer[0]);
 
         $container.empty();
 
+        var $templateWrapper = $("<div>").addClass(DROP_DOWN_EDITOR_FIELD_TEMPLATE_WRAPPER_CLASS).appendTo($container);
```
Comment (churkin) [inline js/ui/drop_down_editor/ui.drop_down_editor.js]: `const`

File: js/ui/text_box/action_button_collection/index.js [hunk 12]
```diff
@@ -61,11 +137,7 @@ export default class ActionButtonCollection {
     getButton(buttonName) {
         const button = find(this.buttons, ({ name }) => name === buttonName);
 
-        if(!button) {
-            throw "Cannot find button with this name";
-        }
-
-        return button.instance;
+        return button ? button.instance : null;
```
Comment (churkin) [inline js/ui/text_box/action_button_collection/index.js:133]: `return button && button.instance;`

File: js/ui/text_box/action_button_collection/index.js [hunk 13]
```diff
@@ -10,46 +63,69 @@ export default class ActionButtonCollection {
...
-        if(!buttons) {
-            buttons = this.defaultButtonsInfo.map(({ name }) => name);
-        }
+    _renderButtons(buttons, $container, targetLocation) {
+        let $buttonsContainer = null;
+        const buttonsInfo = buttons ? this._compileButtonInfo(buttons) : this.defaultButtonsInfo;
+        const getButtonsContainer = () => {
+            $buttonsContainer = $buttonsContainer || $("<div>")
+                .addClass(TEXTEDITOR_BUTTONS_CONTAINER_CLASS)
+                .addClass(targetLocation === "after" ? TEXTEDITOR_AFTER_BUTTONS_CONTAINER_CLASS : TEXTEDITOR_BEFORE_BUTTONS_CONTAINER_CLASS)
```
Comment (churkin) [inline js/ui/text_box/action_button_collection/index.js]: remove extra classes

File: js/ui/text_box/ui.text_editor.base.js [hunk 14]
```diff
@@ -54,37 +55,68 @@ var CONTROL_KEYS = [
...
+            /**
+            * @name dxActionButton.name
+            * @type string
+            * @default undefined
+            */
+            /**
+            /**
+            * @name dxActionButton.location
+            * @type Enums.ActionButtonLocation
+            * @default 'after'
```
Comment (churkin) [inline js/ui/text_box/ui.text_editor.base.js]: `"..."`

File: js/ui/text_box/ui.text_editor.base.js [hunk 15]
```diff
@@ -389,19 +417,24 @@ var TextEditorBase = Editor.inherit({
...
-        $testEditorContainer.append(this._createInput());
-        this.$afterButtonsContainer = this._buttonCollection.renderAfterButtons(buttons, $testEditorContainer);
+        this._$beforeButtonsContainer = this._buttonCollection.renderBeforeButtons(buttons, $textEditorContainer);
+
+        this._$textEditorInputContainer = $("<div>")
+            .addClass(TEXTEDITOR_INPUT_CONTAINER_CLASS)
+            .appendTo($textEditorContainer);
+        this._$textEditorInputContainer.append(this._createInput());
+        this._$afterButtonsContainer = this._buttonCollection.renderAfterButtons(buttons, $textEditorContainer);
+        this._updateButtonsStyling();
```
Comment (churkin) [inline js/ui/text_box/ui.text_editor.base.js]: override `_renderStylingMode`

**Review by churkin — CHANGES_REQUESTED** [review]

File: js/ui/text_box/action_button_collection/button.js (hunk 3 above)
Comment (alexander-kotov-dx) [reply js/ui/text_box/action_button_collection/button.js:4]: The default value is unnecessary in this place. We need location value before the button creation to choose correct buttons container. We apply default location in _renderButtons

File: js/ui/text_box/action_button_collection/index.js (hunk 5 above)
Comment (alexander-kotov-dx) [reply js/ui/text_box/action_button_collection/index.js:11]: Yes, but we going to implement it in the separate PR

File: js/ui/text_box/action_button_collection/index.js (hunk 12 above)
Comment (alexander-kotov-dx) [reply js/ui/text_box/action_button_collection/index.js:133]: i think "null" is better then "undefined" in this case

File: js/ui/drop_down_editor/ui.drop_down_editor.js (hunk 2 above)
Comment (alexander-kotov-dx) [reply js/ui/drop_down_editor/ui.drop_down_editor.js:380]: no, remove can clear buttons content

**Review by DokaRus — APPROVED** [review]

**Review by churkin — APPROVED** [review]

**Pull request merged**